### PR TITLE
feature: Add Patch3D operator

### PR DESCRIPTION
**Pull request by mrava87**
## Description
This PR extends the `pylops.signalprocessing.Patch2D` operator to three-dimensional dataset.

## Definition of done
- Correctly handle edges
- Add example
- Add test

### CODE REVIEW

**Comment by mrava87** [comment]
@cako this is ready. It is a lot of code but I am not sure there is a better way to handle the various edges and corners. Have a look and let me know what you think :)

**Comment by cako** [comment]
I'll have a look. Just a quick (and maybe dumb question), what is the difference between the Patch and Sliding operators?

**Comment by mrava87** [comment]
Thanks! No dumb at at, Sliding3D takes 1 full axis and windows over the other 2. Patch3D does take 3d windows all over the data. 

Just one practical example. For transforms like Radon that are almost invariant to the time axis (apart from edge effects) you can simply use the first, but for things like FKK that are not invariant to time axis you want to have small 3D windows :) makes sense?

File: pylops/utils/tapers.py [hunk 1]
```diff
@@ -210,3 +210,43 @@ def taper3d(nt, nmask, ntap, tapertype="hanning"):
     tpr_3d = np.tile(tpr_yx[:, :, np.newaxis], (1, nt))
 
     return tpr_3d
+
+
+def tapernd(nmask, ntap, tapertype="hanning"):
```
Comment (cako) [inline pylops/utils/tapers.py:215]: Needs to be added to the docs

**Review by cako — CHANGES_REQUESTED** [review]

File: examples/plot_patching.py [hunk 2]
```diff
@@ -45,19 +47,18 @@
 # orthogonal operators, this must be replaced by an inverse.
 nwins = (13, 6)
 nwin = (20, 34)
-nop = (128, 128)
+nop = (128, 128 // 2 + 1)  # second axis is real
```
[the remark this answers is not in the document]
Comment (mrava87) [reply examples/plot_patching.py:50]: Good point, done

File: examples/plot_patching.py [hunk 3]
```diff
@@ -94,11 +92,115 @@
...
+axs[1].set_xlabel(r"$y(m)$")
+plt.tight_layout()
+
+###############################################################################
+# Let's create now the :py:class:`pylops.signalprocessing.Patch3D` operator
+# applying the adjoint of the :py:class:`pylops.signalprocessing.FFT3D`
+# operator to each patch.
+nwins = (5, 4, 3)
+nwin = (20, 20, 34)
+nop = (128, 128, 128 // 2 + 1)  # second axis is real
```
[the remark this answers is not in the document]
Comment (mrava87) [reply examples/plot_patching.py:165]: Done

File: examples/plot_patching.py [hunk 4]
```diff
@@ -94,11 +92,115 @@
...
+    Op.H, dims, dimsd, nwin, nover, nop, tapertype=None
+)
+fftdata = Slid.H * data
+
+Slid = pylops.signalprocessing.Patch3D(
+    Op.H, dims, dimsd, nwin, nover, nop, tapertype="hanning", design=False
+)
+reconstructed_data = np.real(Slid * fftdata)
+
+fig, axs = plt.subplots(1, 2, figsize=(7, 5), sharey=True)
```
[the remark this answers is not in the document]
Comment (mrava87) [reply examples/plot_patching.py:183]: Sure, makes sense, done now :)

File: pylops/signalprocessing/Patch3D.py [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    This operator can be used to perform local, overlapping transforms (e.g.,
+    :obj:`pylops.signalprocessing.FFTND`
+    or :obj:`pylops.signalprocessing.Radon3D`) on 3-dimensional arrays.
+
+    .. note:: The shape of the model has to be consistent with
+       the number of windows for this operator not to return an error. As the
+       number of windows depends directly on the choice of ``nwin`` and
+       ``nover``, it is recommended to use ``design=True`` if unsure about the
+       choice ``dims`` and use the number of windows printed on screen to
+       define such input parameter.
```
[the remark this answers is not in the document]
Comment (mrava87) [reply pylops/signalprocessing/Patch3D.py:44]: I think it makes lots of sense, I also always fail the first time and then use design=True to guide me. I am not just so sure how to do this as we dont really have any operator where things are not built directly when making the operator... maybe we could make a static method that can be called directly (prior to making the operator itself). What about you make a github issue so we dont forget and I can handle this on a separate PR?

File: pylops/utils/tapers.py (hunk 1 above)
Comment (mrava87) [reply pylops/utils/tapers.py:215]: True!

File: pylops/signalprocessing/Patch3D.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cako) [reply pylops/signalprocessing/Patch3D.py:44]: Sure, let's do that!

**Review by cako — APPROVED** [review]

**Pull request merged**